feat: add upgrade access button to user settings page

babblebey opened the pull request:
## Description

<!-- 
Please do not leave this blank 
This PR [adds/removes/fixes/replaces] the [feature/bug/etc]. 
-->
This Pull Request adds the `StripeCheckoutButton` component to the User Settings page.

## What type of PR is this? (check all applicable)

- [x] 🍕 Feature
- [ ] 🐛 Bug Fix
- [ ] 📝 Documentation Update
- [ ] 🎨 Style
- [ ] 🧑‍💻 Code Refactor
- [ ] 🔥 Performance Improvements
- [ ] ✅ Test
- [ ] 🤖 Build
- [ ] 🔁 CI
- [ ] 📦 Chore (Release)
- [ ] ⏩ Revert

## Related Tickets & Documents
<!-- 
Please use this format link issue numbers: Fixes #123
https://docs.github.com/en/free-pro-team@latest/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword 
-->
Fixes #1394 

## Mobile & Desktop Screenshots/Recordings

<!-- Visual changes require screenshots -->
[screencast-localhost_3000-2023.07.20-20_43_11.webm](https://github.com/open-sauced/insights/assets/25631971/28168090-37f5-4129-94d1-0386bc4a4334)

## Added tests?

- [ ] 👍 yes
- [x] 🙅 no, because they aren't needed
- [ ] 🙋 no, because I need help

## Added to documentation?

- [ ] 📜 README.md
- [ ] 📓 docs.opensauced.pizza
- [ ] 🍕 dev.to/opensauced
- [ ] 📕 storybook
- [x] 🙅 no documentation needed

## [optional] Are there any post-deployment tasks we need to perform?

NA

## [optional] What gif best describes this PR or how it makes you feel?



<!-- note: PRs with deleted sections will be marked invalid -->

<!--
  For Work In Progress Pull Requests, please use the Draft PR feature,
  see https://github.blog/2019-02-14-introducing-draft-pull-requests/ for further details.
  
  For a timely review/response, please avoid force-pushing additional
  commits if your PR already received reviews or comments.
  
  Before submitting a Pull Request, please ensure you've done the following:
  - 📖 Read the Open Sauced Contributing Guide: https://github.com/open-sauced/.github/blob/main/CONTRIBUTING.md.
  - 📖 Read the Open Sauced Code of Conduct: https://github.com/open-sauced/.github/blob/main/CODE_OF_CONDUCT.md.
  - 👷‍♀️ Create small PRs. In most cases, this will be possible.
  - ✅ Provide tests for your changes.
  - 📝 Use descriptive commit messages.
  - 📗 Update any related documentation and include any relevant screenshots.
-->

brandonroberts commented on components/organisms/UserSettingsPage/user-settings-page.tsx:
```diff
@@ -371,6 +372,12 @@ const UserSettingsPage = ({ user }: userSettingsPageProps) => {
           </div>
+          <div className="flex flex-col gap-3 order-first md:order-last">
+            <label className="text-2xl font-normal text-light-slate-11">Upgrade Access</label>
+            <div className="text-center p-3 border border-dashed rounded-xl border-light-slate-8">
```
I don't think it needs a border around the box, and if the button can be left aligned to be consistent with the other ones.

babblebey replied:
Sure.

I will go on and refactor the `StripeCheckoutButton` component to be more customisable for usage in this context. Its current set-up is hard coded to work this way as integrated with the Dashboard's `Reports` component.

bdougie replied:
Making the component more flexible would be ideal in this case.

babblebey replied:
Done 🍕

NsdHSO commented on components/organisms/StripeCheckoutButton/stripe-checkout-button.tsx:
```diff
@@ -32,17 +34,11 @@ const StripeCheckoutButton = () => {
-    </div>
+    <form onSubmit={handleFormSubmit}>
+      <Button {...props}>
+        { children || "Upgrade Access" }
```
```suggestion
        {children || "Upgrade Access"}
```

NsdHSO commented on components/organisms/UserSettingsPage/user-settings-page.tsx:
```diff
@@ -371,6 +373,15 @@ const UserSettingsPage = ({ user }: userSettingsPageProps) => {
+          <div className="flex flex-col gap-6 order-first md:order-last">
+            <div className="flex flex-col gap-3">
+              <label className="text-2xl font-normal text-light-slate-11">Upgrade Access</label>
+              <div className="w-full sm:max-w-[330px]">
```
```suggestion
              <div className="w-full sm:max-w-80]">
```

babblebey commented:
Changes Made to make the `StripeCheckoutButton` component more flexible:

- Modified the `StripeCheckoutButton` component to accept `children` and `props` of type `StripeCheckoutButtonProps`. The `children` prop allows users to customize the text content of the button, while the `props` can be used to pass additional styling or behavior options.
- Removed the unnecessary `<div>` element and replaced it with a `<form>` element to wrap the `Button`. This ensures better semantic structure for form submission.
- Removed the redundant `<p>` element around the `<form>` element, as it is no longer necessary.
- Updated the `Button` component usage to spread the `props` object so that any additional props provided to `StripeCheckoutButton` can be passed down to the `Button` component.

Overall, these changes have made the `StripeCheckoutButton` component more flexible and reusable, allowing users to easily customize the button's appearance and behavior based on their specific use cases. _Sample Use Cases seen below_

**The dashboard `Reports` Component use case**

```javascript
<>
  <Text>Upgrade to a subscription to gain access to generate custom reports!</Text>
  <p className="flex justify-center py-4 px-2">
    // HERE
    <StripeCheckoutButton variant="primary" className="w-52 h-[38px] flex justify-center"/>
  </p>
</>
```

![screenshot-localhost_3000-2023 07 21-11_30_31](https://github.com/open-sauced/insights/assets/25631971/6b2e9ef7-38d3-4062-b521-062273dbc63f)

**The User Settings page use case**

```javascript
<div className="flex flex-col gap-6 order-first md:order-last">
  <div className="flex flex-col gap-3">
    <label className="text-2xl font-normal text-light-slate-11">Upgrade Access</label>
    <div className="w-full sm:max-w-[330px]">
      <Text>Upgrade to a subscription to gain access to generate custom reports!</Text>
    </div>
  </div>
  // HERE
  <StripeCheckoutButton variant="primary" />
</div>
```

![screenshot-localhost_3000-2023 07 21-12_50_04](https://github.com/open-sauced/insights/assets/25631971/5e0b8b07-87fa-4ba4-b7ab-6ad7f9407400)

**The User Settings page use case (with custom button content)**
```javascript
<StripeCheckoutButton variant="primary">
  Subscribe Now
</StripeCheckoutButton>
```

![screenshot-localhost_3000-2023 07 21-12_51_39](https://github.com/open-sauced/insights/assets/25631971/7bcd5a7b-f37b-47b2-a0ae-ea6fc99b6a5c)

brandonroberts commented:
@babblebey changes look better. Got some build/test failures on the PR preview

babblebey commented:
> changes look better. Got some build/test failures on the PR preview

I was out of sync with the new build configuration #1396

Now Fixed 🍕

brandonroberts commented:
@babblebey thanks. One more addition is that button shouldn't be displayed if the user account already has access. The `hasReports` property from the `useSession` hook gives you this info https://github.com/open-sauced/insights/blob/beta/lib/hooks/useSession.ts#L43

Should be good to go then

babblebey commented:
> thanks. One more addition is that button shouldn't be displayed if the user account already has access. 

Haha, funny how I didn't think about that logic 

Will do 🍕
